File: app/src/main/java/de/xikolo/controllers/course/CertificatesFragment.java
package de.xikolo.controllers.course;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.NestedScrollView;
import android.text.Html;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.yatatsu.autobundle.AutoBundleField;

import java.io.File;
import java.io.IOException;

import butterknife.BindView;
import de.xikolo.App;
import de.xikolo.R;
import de.xikolo.controllers.base.LoadingStatePresenterFragment;
import de.xikolo.models.Course;
import de.xikolo.presenters.base.PresenterFactory;
import de.xikolo.presenters.course.CertificatesPresenter;
import de.xikolo.presenters.course.CertificatesPresenterFactory;
import de.xikolo.presenters.course.CertificatesView;
import de.xikolo.services.DownloadService;
import de.xikolo.utils.IntentUtil;

public class CertificatesFragment extends LoadingStatePresenterFragment<CertificatesPresenter, CertificatesView> implements CertificatesView {

    public final static String TAG = CertificatesFragment.class.getSimpleName();

    @AutoBundleField String courseId;

    @BindView(R.id.content_view) NestedScrollView scrollView;
    @BindView(R.id.container) LinearLayout container;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public int getLayoutResource() {
        return R.layout.content_certificates;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void showCertificates(Course course) {
        boolean isEnrolled = course.isEnrolled();

        container.removeAllViews();

        if (course.certificates.qualifiedCertificateAvailable)
            setupItem(getString(R.string.course_qualified_certificate),
                    getString(R.string.course_qualified_certificate_desc),
                    isEnrolled,
                    course.certificates.qualifiedCertificateUrl != null,
                    v -> IntentUtil.openDoc(App.getInstance(), course.certificates.qualifiedCertificateUrl)
            );

        if (course.certificates.recordOfAchievementAvailable)
            setupItem(getString(R.string.course_record_of_achievement),
                    getString(R.string.course_record_of_achievement_desc, course.certificates.recordOfAchievementThreshold),
                    isEnrolled,
                    course.certificates.recordOfAchievementUrl != null,
                    v -> IntentUtil.openDoc(App.getInstance(), course.certificates.recordOfAchievementUrl)
            );

        if (course.certificates.confirmationOfParticipationAvailable)
            setupItem(getString(R.string.course_confirmation_of_participation),
                    getString(R.string.course_confirmation_of_participation_desc, course.certificates.confirmationOfParticipationThreshold),
                    isEnrolled,
                    course.certificates.confirmationOfParticipationUrl != null,
                    (v) -> {
                        Intent intent = new Intent(App.getInstance(), DownloadService.class);
                        Bundle bundle = new Bundle();
                        bundle.putString(DownloadService.ARG_TITLE, getString(R.string.course_confirmation_of_participation));
                        bundle.putString(DownloadService.ARG_URL, course.certificates.confirmationOfParticipationUrl);
                        try {
                            File path = File.createTempFile(course.id + "_confirmationofparticipation", "pdf");
                            path.delete();

                            bundle.putString(DownloadService.ARG_FILE_PATH, path.getAbsolutePath());

                            intent.putExtras(bundle);
                            App.getInstance().startService(intent);
                            while (!path.exists()) {

                            }
                            IntentUtil.openDoc(App.getInstance(), path.getAbsolutePath());
                        } catch (IOException e) {
                            e.printStackTrace();
                        }

                    });
    }

    private void setupItem(String header, String text, boolean enrolled, boolean documentAvailable, View.OnClickListener downloadClickListener) {
        View h = getLayoutInflater().inflate(R.layout.item_section_header, null);
        ((TextView) h.findViewById(R.id.textHeader)).setText(header);

        View v = getLayoutInflater().inflate(R.layout.item_certificate, null);
        ((TextView) v.findViewById(R.id.textContent)).setText(Html.fromHtml(text));
        Button downloadButton = v.findViewById(R.id.button_certificate_download);
        if (documentAvailable) {
            downloadButton.setText(R.string.course_certificate_view);
            downloadButton.setOnClickListener(downloadClickListener);
        } else {
            if (enrolled) {
                downloadButton.setEnabled(false);
                downloadButton.setText(R.string.course_certificate_not_available);
            } else {
                downloadButton.setVisibility(View.GONE);
            }
        }

        container.addView(h);
        container.addView(v);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.refresh, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        switch (itemId) {
            case R.id.action_refresh:
                onRefresh();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @NonNull
    @Override
    protected PresenterFactory<CertificatesPresenter> getPresenterFactory() {
        return new CertificatesPresenterFactory(courseId);
    }

}

File: app/src/main/java/de/xikolo/controllers/course/CertificatesFragment.kt
package de.xikolo.controllers.course

import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import androidx.core.widget.NestedScrollView
import androidx.core.widget.TextViewCompat
import butterknife.BindView
import com.yatatsu.autobundle.AutoBundleField
import de.xikolo.R
import de.xikolo.controllers.base.ViewModelFragment
import de.xikolo.controllers.helper.DownloadViewHelper
import de.xikolo.extensions.observe
import de.xikolo.models.Course
import de.xikolo.models.DownloadAsset
import de.xikolo.models.Enrollment
import de.xikolo.models.dao.EnrollmentDao
import de.xikolo.viewmodels.course.CertificateListViewModel

class CertificatesFragment : ViewModelFragment<CertificateListViewModel>() {

    companion object {
        val TAG: String = CertificatesFragment::class.java.simpleName
    }

    @AutoBundleField
    lateinit var courseId: String

    @BindView(R.id.content_view)
    internal lateinit var scrollView: NestedScrollView

    @BindView(R.id.container)
    internal lateinit var container: LinearLayout

    override val layoutResource = R.layout.fragment_course_certificates

    override fun createViewModel(): CertificateListViewModel {
        return CertificateListViewModel(courseId)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.course
            .observe(viewLifecycleOwner) {
                showCertificates(
                    it,
                    EnrollmentDao.Unmanaged.findForCourse(it.id)
                )
            }
    }

    private fun showCertificates(course: Course, enrollment: Enrollment?) {
        container.removeAllViews()

        activity?.let { activity ->
            val downloadViewHelpers: MutableList<DownloadViewHelper> = ArrayList()

            if (course.certificates.confirmationOfParticipation.available) {
                val dvh = DownloadViewHelper(
                    activity,
                    DownloadAsset.Certificate.ConfirmationOfParticipation(
                        enrollment?.certificates?.confirmationOfParticipationUrl,
                        course
                    ),
                    getString(R.string.course_confirmation_of_participation),
                    String.format(getString(R.string.course_confirmation_of_participation_desc), course.certificates.confirmationOfParticipation.threshold),
                    getString(R.string.course_certificate_not_achieved)
                )

                downloadViewHelpers.add(dvh)
            }

            if (course.certificates.recordOfAchievement.available) {
                val dvh = DownloadViewHelper(
                    activity,
                    DownloadAsset.Certificate.RecordOfAchievement(enrollment?.certificates?.recordOfAchievementUrl, course),
                    getString(R.string.course_record_of_achievement),
                    String.format(getString(R.string.course_record_of_achievement_desc), course.certificates.recordOfAchievement.threshold),
                    getString(R.string.course_certificate_not_achieved)
                )

                downloadViewHelpers.add(dvh)
            }

            if (course.certificates.qualifiedCertificate.available) {
                val dvh = DownloadViewHelper(
                    activity,
                    DownloadAsset.Certificate.QualifiedCertificate(enrollment?.certificates?.qualifiedCertificateUrl, course),
                    getString(R.string.course_qualified_certificate),
                    getString(R.string.course_qualified_certificate_desc),
                    getString(R.string.course_certificate_not_achieved)
                )

                downloadViewHelpers.add(dvh)
            }

            downloadViewHelpers.forEach { dvh ->
                TextViewCompat.setTextAppearance(dvh.textFileName, R.style.TextAppearanceMedium)
                if (!course.isEnrolled) {
                    dvh.buttonDownloadStart.visibility = View.GONE
                }
                container.addView(dvh.view)
            }

            if (downloadViewHelpers.isEmpty()) {
                showEmptyMessage(R.string.empty_message_course_certificates)
            } else {
                showContent()
            }
        }
    }

}
